### PR TITLE
Working FgFlex implmentation

### PR DESCRIPTION
**Pull request by pmhalvor**
FgFlex works, but is a simplified aggregation of RACL and IMN.

Results so far give hard metrics of 0, but fine-tuning in process to find some semi-optimal setup

### CODE REVIEW

File: src/model.py [hunk 1]
```diff
@@ -1734,3 +1844,320 @@ def get_confidence(expression_logits, batch, current_epoch):
...
+        Borrowed directly from https://github.com/ruidan/IMN-E2E-ABSA
+        
+        Compute the probability of using gold opinion labels in opinion transmission
+        (To alleviate the problem of unreliable predictions of opinion labels sent from AE to AS,
+        in the early stage of training, we use gold labels as prediction with probability 
+        that depends on the number of current epoch)
+
+        """
+        prob = constant/(constant+torch.exp(torch.tensor(epoch_count).float()/constant))
+        return prob
```
Comment (pmhalvor) [inline src/model.py:2146]: Percentage of how much influence true labels should have on attention inputs, according to current epoch count. Borrowed from IMN codebase, w/ the added flexibility to configure the constant used, via `warm_up_constant` parameter in json-configs.

File: src/model.py [hunk 2]
```diff
@@ -1734,3 +1844,320 @@ def get_confidence(expression_logits, batch, current_epoch):
...
+        # decide how much true labels should influence attention based on current epoch
+        gold_influence = self.get_prob(self.current_epoch, self.find("warm_up_constant", default=5))
+
+        focus_scope = (  # strengthen signal from true logits
+            gold_influence*true.bool().float()
+                + (1-gold_influence)*logits.argmax(-1)
+        ).detach().to(torch.device(self.device))
+        focus_scope.requires_grad = True
+        
+        return focus_scope  # shape: [batch, sequence]
```
Comment (pmhalvor) [inline src/model.py:2162]: Related to `get_prob()`, strengthens signal to tokens deemed important according to true labels and models preditions.

File: src/study.py [hunk 3]
```diff
@@ -335,7 +337,12 @@ def save_model(self):
                                 print("FAILED TO SAVE MODEL {} DUE TO FOLLOW EXCEPTION".format(study.name))
                                 print(e)
         
+                    # free up memory again
+                    metric = str(study.metric)
+                    del study.model
+                    torch.cuda.empty_cache()
```
Comment (pmhalvor) [inline src/study.py:343]: No need to hold on to models after study is complete. Was sometimes causing OOM errors on slurm-based servers. This ensures to free up some of that space again to protect against unneeded failures.

**Review by pmhalvor — COMMENTED** [review]
Ready for merge. 

See comments for architectural decisions, including areas for future development. 
Main changes done in `model.py`
